R/cube-subset.R

```diff
@@ -103,6 +128,34 @@ setMethod("[", "CrunchCube", function(x, i, j, ..., drop = TRUE) {
     return(out)
 })
 
+#' Transform character vectors into indices
+#'
+#' When a user supplies character strings to subset a cube, we need to translate it
+#' into a numeric index to subset the cube. This function does that as well as
+#' some error checking. It is broken out for testing.
+#'
+#' @param cat_names The category names for a given dimension
+#' @param idx The index
+#' @param visible whether or not a category is visible to the user
+#'
+#' @keywords internal
+replaceCharWithNumeric <- function(cat_names, idx, visible = TRUE) {
+    if (!is.logical(idx) &&
+        length(idx) != length(unique(idx))) {
+        halt("Index is not unique. Cube subetting is only supported for unique indices.")
+    }
+
+    if (is.character(idx)) {
+        visible_categories <- cat_names[visible]
+        not_categories <- !(idx %in% visible_categories)
+        if (any(not_categories)) {
+            halt("Invalid categories: ", serialPaste(idx[not_categories]))
+        }
+        return(match(idx, visible_categories))
```

Comment:
Although it _shouldn't_ be the case there could be edge cases where there is more than one of the same name (currently insertions+category names aren't required to be unique). I wonder if `which(idx %in% visible_categories)` would be safer here in case we run into that particular edge.